test/fixtures/gcs-fixture/src/main/java/fixture/gcs/GoogleCloudStorageHttpHandler.java

```diff
@@ -239,50 +239,55 @@ public void handle(final HttpExchange exchange) throws IOException {
                 final Map<String, String> params = new HashMap<>();
                 RestUtils.decodeQueryString(exchange.getRequestURI().getQuery(), 0, params);
 
-                final String blobName = params.get("test_blob_name");
-                if (blobs.containsKey(blobName) == false) {
-                    exchange.sendResponseHeaders(RestStatus.NOT_FOUND.getStatus(), -1);
-                    return;
+                final String range = requireHeader(exchange, "Content-Range");
+                final HttpHeaderParser.ContentRange contentRange = HttpHeaderParser.parseContentRangeHeader(range);
+                if (contentRange == null) {
+                    throw failAndThrow("Invalid Content-Range: " + range);
                 }
-                BytesReference blob = blobs.get(blobName);
-                final String range = exchange.getRequestHeaders().getFirst("Content-Range");
-                final Integer limit = getContentRangeLimit(range);
-
-                blob = CompositeBytesReference.of(blob, requestBody);
-                blobs.put(blobName, blob);
-
-                if (limit == null) {
-                    if ("bytes */*".equals(range) == false) {
-                        final int start = getContentRangeStart(range);
-                        final int end = getContentRangeEnd(range);
-                        exchange.getResponseHeaders().add("Range", String.format(Locale.ROOT, "bytes=%d-%d", start, end));
-                    }
-                    exchange.getResponseHeaders().add("Content-Length", "0");
-                    exchange.sendResponseHeaders(308 /* Resume Incomplete */, -1);
-                } else {
-                    if (limit > blob.length()) {
-                        throw new AssertionError("Requesting more bytes than available for blob");
-                    }
-                    exchange.sendResponseHeaders(RestStatus.OK.getStatus(), -1);
+
+                final MockGcsBlobStore.UpdateResponse updateResponse = mockGcsBlobStore.updateResumableUpload(
+                    params.get("upload_id"),
+                    contentRange,
+                    requestBody
+                );
+
+                if (updateResponse.rangeHeader() != null) {
+                    exchange.getResponseHeaders().add("Range", range);
                 }
+                exchange.getResponseHeaders().add("Content-Length", "0");
+                exchange.sendResponseHeaders(updateResponse.statusCode(), -1);
             } else {
                 exchange.sendResponseHeaders(RestStatus.NOT_FOUND.getStatus(), -1);
             }
+        } catch (MockGcsBlobStore.GcsRestException e) {
+            sendError(exchange, e);
         } finally {
-            int read = exchange.getRequestBody().read();
-            assert read == -1 : "Request body should have been fully read here but saw [" + read + "]";
```

Comment:
This seemed to throw `IOExceptions` due to the body being closed, `ExchangeImpl` calls `#close()` when you call `sendResponseHeaders` with a zero-length body. None of the other `HttpHandler`s had this check.